bluefog/torch/optimizers.py

```diff
@@ -462,6 +463,343 @@ def step(self, closure=None):
+                    "should be parameter-wise tep func:\n"
+                    "   func(parameter, gradient, parameter_group) -> None\n")
+            self._step_func(p, grad.data, param_group)
+
+            if self._reduce_delay[p] <= 0:
+                raise AssertionError(
+                    "Unexpected behavior: forward computation were computed "
+                    "more than num_steps_per_communication times before call "
+                    "to step(). Adjust num_steps_per_communication to "
+                    "accumulate gradients locally.")
+            self._reduce_delay[p] -= 1
+            if self._reduce_delay[p] == 0:
+                if self._communication_type == CommunicationType.allreduce:
+                    handle = self._allreduce_data_async(p)
+                elif self._communication_type == CommunicationType.neighbor_allreduce:
+                    handle = self._neighbor_allreduce_data_async(p)
+                elif self._communication_type == CommunicationType.hierarchical_neighbor_allreduce:
+                    handle = self._hierarchical_neighbor_allreduce_data_async(p)
+                elif self._communication_type == CommunicationType.empty:
+                    handle = None
+                else:
+                    raise ValueError("Unsuppported CommunicationType encountered.")
+                self._handles[p] = handle
+
+        return hook
+
+    def _sgd_step(self, p, grad, param_group):
+        """Parameter-wise of torch.optim.SGD.step."""
+        weight_decay = param_group['weight_decay']
+        momentum = param_group['momentum']
+        dampening = param_group['dampening']
+        nesterov = param_group['nesterov']
+        lr = param_group['lr']
+        d_p = grad
+        if weight_decay != 0:
+            d_p.add_(p.data, alpha=weight_decay)
+        if momentum != 0:
+            param_state = self.state[p]
+            if 'momentum_buffer' not in param_state:
+                buf = param_state['momentum_buffer'] = torch.clone(d_p).detach()
+            else:
+                buf = param_state['momentum_buffer']
+                buf.mul_(momentum).add_(d_p, alpha=1 - dampening)
+            if nesterov:
+                d_p = d_p.add(buf, alpha=momentum)
+            else:
+                d_p = buf
+        p.data.add_(d_p, alpha=-lr)
+
+    def _adam_step(self, p, grad, param_group):
+        """Parameter-wise of torch.optim.Adam.step."""
+        if grad.is_sparse:
+            raise RuntimeError(
+                'Adam does not support sparse gradients, please consider SparseAdam instead')
+        amsgrad = param_group['amsgrad']
+        state = self.state[p]
+
+        # State initialization
+        if not state:
+            state['step'] = 0
+            # Exponential moving average of gradient values
+            state['exp_avg'] = torch.zeros_like(p.data, memory_format=torch.preserve_format)
```

Comment:
Seems that memory_format is not supported by PyTorch before 1.5.0? Do we need to change here or the requirements.txt?

Reply:
Good point. Yes, I copied the latest pytorch. Didn't realize this problem. Let me modify it.